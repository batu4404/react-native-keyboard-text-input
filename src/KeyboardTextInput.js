import React, { Component } from 'react';
import {
    Text,
    View,
    TextInput,
    StyleSheet,
    KeyboardAvoidingView,
    Keyboard,
    DeviceEventEmitter,
    Modal,
    Dimensions,
    TouchableWithoutFeedback,
} from 'react-native';
import PropsType from 'prop-types';
import { setTimeout } from 'core-js/library/web/timers';
import { getPropsWithoutFunction } from './utils';

export default class KeyboardTextInput extends Component {
    constructor(props) {
        super(props);

        let propsWithoutFunction = getPropsWithoutFunction(this.props);

        this.state = {
            value: this.props.value,
            modalVisible: false,
            propsWithoutFunction,
        }
    }

    componentWillReceiveProps (nextProps) {
        let propsWithoutFunction = getPropsWithoutFunction(this.props);
        let value = nextProps.value;

        this.setState({value, propsWithoutFunction});
    }

    componentDidUpdate(nextProps, nextState) {
        if (nextState.modalVisible) {
            this.registerKeyboardEvent();
        }
    }

    registerKeyboardEvent = () => {
        this.keyboardDidHideListener = Keyboard.addListener('keyboardDidHide', this._keyboardDidHide);
    }

    unregisterKeyboardEvent = () => {
        this.keyboardDidHideListener.remove();
    }
    
    _keyboardDidHide = (e) => {
        if (this.state.modalVisible) {
            // to fix "trying to add local data to unkown tag ..." (haven't known reason)
            setTimeout(() => {
                this.onEndEditing();
            }, 20);
        }
    }

    onPress = () => {
        this.setState({modalVisible: true});
    }

    onFocus = () => {
        this.registerKeyboardEvent();
    }

    onEndEditing = () => {
        this.unregisterKeyboardEvent();
        this.setState({modalVisible: false});

        this.props.onEndEditing(this.state.value);
    }

    onChangeText = (text) => {
        this.setState({
            value: text,
        });
        this.props.onChangeText(text);
    }

    handleTouchOnSpace = () => {
        this.onEndEditing();
    }

    renderModal() {
        const {
            style,
            value,
            editable,
            ...restProps
        } = this.state.propsWithoutFunction;

        return (
            <Modal
                onRequestClose={() => {this.props.onClose()}}
                visible={this.state.modalVisible}
                transparent={true}
                animationType='slide'
            >      
                <View style={styles.modalContainer}>
                    <TouchableWithoutFeedback
                        onPress={this.handleTouchOnSpace}
                    >
                        <View style={styles.flexOne} />
                    </TouchableWithoutFeedback>

                    <View>
                        <TextInput
                            placeholder={this.props.placeholder}
                            style={[styles.keyboardInput, this.props.inputStyle]}
                            onFocus={this.onFocus}
                            value={this.state.value}
                            autoFocus={this.state.modalVisible}
                            onEndEditing={this.onEndEditing}
                            onChangeText={this.onChangeText}
                            underlineColorAndroid='transparent'
                            {...restProps}
                        />
                    </View>
                </View>
            </Modal>
        )
    }

    render() {	
        const {
            style,
            value,
            editable,
            ...restProps
        } = this.state.propsWithoutFunction;

        return (
            <View>
                {this.renderModal()}

                <TextInput 
                    value={this.state.value}
                    style={[this.props.style, editable ? styles.editable : styles.readOnly, style]}
                    numberOfLines={1}
                    placeholder={this.props.placeholder}
                    underlineColorAndroid='transparent'
                    editable={false}
                    {...restProps}
                />
                <View
                    onPress={this.onPress}
                    style={{
                        position: 'absolute',
                        top: 0,
                        left: 0,
                        bottom: 0,
                        right: 0,
                    }}
                >
                    <TouchableWithoutFeedback 
                        onPress={this.onPress}
                    >
                        <View style={styles.flexOne} />
                    </TouchableWithoutFeedback>
                </View> 
            </View>
        )
    }
}

KeyboardTextInput.propTypes = {
    value: PropsType.string,
    keyboardType: PropsType.string,
    editable: PropsType.bool,
    onEndEditing: PropsType.func,
    onChangeText: PropsType.func,
}

KeyboardTextInput.defaultProps = {
    value: undefined,
    keyboardType: 'default',
    editable: true,
    onEndEditing: () => null,
    onChangeText: () => null,
}

const styles = StyleSheet.create({
    modalContainer: {
        flex: 1,
        backgroundColor: 'rgba(0, 0, 0, 0.4)',
    },
    input: {
        backgroundColor: 'white',
        borderWidth: 1,
        borderColor: '#AAA',
        paddingLeft: 10,
    },
    keyboardInput: {
        backgroundColor: 'white',
        borderWidth: 1,
        borderColor: '#AAA',
        paddingLeft: 10,
    },
    editable: {
        backgroundColor: 'white',
    },
    readOnly: {
        backgroundColor: 'whitesmoke',
    },
    flexOne: {
        flex: 1,
    }
});