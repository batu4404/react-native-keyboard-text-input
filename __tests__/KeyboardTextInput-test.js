import React from 'react';
import KeyboardTextInput from '../src/KeyboardTextInput';

import renderer from 'react-test-renderer';

describe('>>> KeyboardTextInput --- Test rendering component', () => {
    it('should render correctly', () => {
        const tree = renderer.create(
            <KeyboardTextInput />
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
