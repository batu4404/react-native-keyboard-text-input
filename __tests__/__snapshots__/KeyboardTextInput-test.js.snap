// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>> KeyboardTextInput --- Test rendering component should render correctly 1`] = `
<View>
  <Modal
    animationType="slide"
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    transparent={true}
    visible={false}
  >
    <View
      style={
        Object {
          "backgroundColor": "rgba(0, 0, 0, 0.4)",
          "flex": 1,
        }
      }
    >
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        hitSlop={undefined}
        nativeID={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "flex": 1,
          }
        }
        testID={undefined}
      />
      <View>
        <TextInput
          allowFontScaling={true}
          autoFocus={false}
          keyboardType="default"
          onChangeText={[Function]}
          onEndEditing={[Function]}
          onFocus={[Function]}
          placeholder={undefined}
          style={
            Array [
              Object {
                "backgroundColor": "white",
                "borderColor": "#AAA",
                "borderWidth": 1,
                "paddingLeft": 10,
              },
              undefined,
            ]
          }
          underlineColorAndroid="transparent"
          value={undefined}
        />
      </View>
    </View>
  </Modal>
  <TextInput
    allowFontScaling={true}
    editable={false}
    keyboardType="default"
    numberOfLines={1}
    placeholder={undefined}
    style={
      Array [
        undefined,
        Object {
          "backgroundColor": "white",
        },
        undefined,
      ]
    }
    underlineColorAndroid="transparent"
    value={undefined}
  />
  <View
    onPress={[Function]}
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      hitSlop={undefined}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "flex": 1,
        }
      }
      testID={undefined}
    />
  </View>
</View>
`;
